File: app/mobile_app/lib/services/post_service.dart
import '../models/api_response.dart';
import '../models/paged_result.dart';
import '../models/post.dart';
import '../models/comment.dart';
import 'api_service.dart';

class PostService {
  final _api = ApiService().dio;

  ApiResponse<PagedResult<PostDto>> _parsePagedPosts(dynamic data) {
    return ApiResponse<PagedResult<PostDto>>.fromJson(
      data as Map<String, dynamic>,
      (json) => PagedResult<PostDto>.fromJson(
        json as Map<String, dynamic>,
        (item) => PostDto.fromJson(item as Map<String, dynamic>),
      ),
    );
  }

  Future<ApiResponse<PagedResult<PostDto>>> getPosts({
    int pageIndex = 1,
    int pageSize = 10,
    String? keyword,
  }) async {
    final response = await _api.get(
      '/Post',
      queryParameters: {
        'PageIndex': pageIndex,
        'PageSize': pageSize,
        'Keyword': keyword,
      }..removeWhere((_, value) => value == null),
    );
    return _parsePagedPosts(response.data);
  }

  Future<ApiResponse<PagedResult<PostDto>>> searchPosts({
    int pageIndex = 1,
    int pageSize = 10,
    String? keyword,
  }) async {
    final response = await _api.get(
      '/Post/search',
      queryParameters: {
        'PageIndex': pageIndex,
        'PageSize': pageSize,
        'Keyword': keyword,
      }..removeWhere((_, value) => value == null),
    );
    return _parsePagedPosts(response.data);
  }

  Future<ApiResponse<PagedResult<PostDto>>> getUserPosts(
    String userId, {
    int pageIndex = 1,
    int pageSize = 10,
    String? keyword,
  }) async {
    final response = await _api.get(
      '/Post/user/$userId',
      queryParameters: {
        'PageIndex': pageIndex,
        'PageSize': pageSize,
        'Keyword': keyword,
      }..removeWhere((_, value) => value == null),
    );
    return _parsePagedPosts(response.data);
  }

  Future<ApiResponse<List<PostDto>>> getHotPosts({int count = 10}) async {
    final response = await _api.get(
      '/Post/hot',
      queryParameters: {'count': count},
    );
    return ApiResponse<List<PostDto>>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => (json as List)
          .map((item) => PostDto.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }

  Future<ApiResponse<PostDto>> createPost(PostCreateDto request) async {
    final response = await _api.post('/Post', data: request.toJson());
    return ApiResponse<PostDto>.fromJson(
      response.data,
      (json) => PostDto.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<ApiResponse<PostDto>> getPost(String id) async {
    final response = await _api.get('/Post/$id');
    return ApiResponse<PostDto>.fromJson(
      response.data,
      (json) => PostDto.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<ApiResponse<PostDto>> updatePost(
    String id,
    PostUpdateDto request,
  ) async {
    final response = await _api.put('/Post/$id', data: request.toJson());
    return ApiResponse<PostDto>.fromJson(
      response.data,
      (json) => PostDto.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<ApiResponse> deletePost(String id) async {
    final response = await _api.delete('/Post/$id');
    return ApiResponse.fromJson(response.data, (json) => json);
  }

  Future<ApiResponse> likePost(String id) async {
    final response = await _api.post('/Post/$id/like');
    return ApiResponse.fromJson(response.data, (json) => json);
  }

  Future<ApiResponse> dislikePost(String id) async {
    final response = await _api.post('/Post/$id/dislike');
    return ApiResponse.fromJson(response.data, (json) => json);
  }

  Future<ApiResponse<List<CommentDto>>> getPostComments(
    String id,
    String type,
  ) async {
    final response = await _api.get('/Post/$id/comments/$type');
    return ApiResponse<List<CommentDto>>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => (json as List)
          .map((i) => CommentDto.fromJson(i as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: app/webapp/src/services/postService.ts
import { apiClient } from './apiClient';
import type {
  PostDto,
  PostCreateDto,
  PostUpdateDto,
  PagedResult,
  CommentDto,
} from '../types';

export const postService = {
  getPosts: (pageIndex?: number, pageSize?: number, keyword?: string) => {
    const params = new URLSearchParams();
    if (pageIndex !== undefined) params.append('PageIndex', pageIndex.toString());
    if (pageSize !== undefined) params.append('PageSize', pageSize.toString());
    if (keyword) params.append('Keyword', keyword);
    return apiClient.get<PagedResult<PostDto>>(`/Post?${params.toString()}`);
  },

  getHotPosts: (count: number = 10) =>
    apiClient.get<PostDto[]>(`/Post/hot?count=${count}`),

  getPost: (id: string, options?: RequestInit) =>
    apiClient.get<PostDto>(`/Post/${id}`, options),

  createPost: (data: PostCreateDto) =>
    apiClient.post<PostDto>('/Post', data),

  updatePost: (id: string, data: PostUpdateDto) =>
    apiClient.put<PostDto>(`/Post/${id}`, data),

  deletePost: (id: string) =>
    apiClient.delete<void>(`/Post/${id}`),

  likePost: (id: string) =>
    apiClient.post<void>(`/Post/${id}/like`),

  dislikePost: (id: string) =>
    apiClient.post<void>(`/Post/${id}/dislike`),

  getPostComments: (id: string, type: string) =>
    apiClient.get<CommentDto[]>(`/Post/${id}/comments/${type}`),

  searchPosts: (pageIndex?: number, pageSize?: number, keyword?: string) => {
    const params = new URLSearchParams();
    if (pageIndex !== undefined) params.append('PageIndex', pageIndex.toString());
    if (pageSize !== undefined) params.append('PageSize', pageSize.toString());
    if (keyword) params.append('Keyword', keyword);
    return apiClient.get<PagedResult<PostDto>>(`/Post/search?${params.toString()}`);
  },
};
